opencl-aot/source/main.cpp

```diff
@@ -0,0 +1,488 @@
+//===------ opencl-aot/source/main.cpp - opencl-aot tool --------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+///
+/// \file
+/// This file contains the implementation of the opencl-aot tool which generates
+/// OpenCL program binary from SPIR-V binary for Intel(R) processor, Intel(R)
+/// Processor Graphics and Intel(R) FPGA Emulation Platform for OpenCL(TM)
+/// devices.
+///
+//===----------------------------------------------------------------------===//
+
+#include "utils.h"
+#include <CL/cl.h>
+
+#include "llvm/Support/CommandLine.h"
+#include "llvm/Support/Path.h"
+
+#include <algorithm>
+#include <cassert>
+#include <fstream>
+#include <iostream>
+#include <map>
+#include <memory>
+#include <string>
+#include <unordered_map>
+#include <vector>
+
+using namespace llvm;
+
+// cl_program, shall initialized with
+auto ProgramDeleter = [](cl_program CLProg) {
+  if (CLProg != nullptr) {
+    cl_int CLErr = clReleaseProgram(CLProg);
+    if (clFailed(CLErr)) {
+      std::cerr << formatCLError("Failed to release OpenCL program", CLErr);
+    }
+  }
+};
+
+using CLProgramUPtr = std::unique_ptr<
+    _cl_program, decltype(ProgramDeleter)>; // decltype(CLProgramUPtr.get()) is
+                                            // same as cl_program, should
+                                            // initialized with ProgramDeleter
+
+auto ContextDeleter = [](cl_context CLCon) {
+  if (CLCon != nullptr) {
+    cl_int CLErr = clReleaseContext(CLCon);
+    if (clFailed(CLErr)) {
+      std::cerr << formatCLError("Failed to release OpenCL context", CLErr);
+    }
+  }
+};
+
+using CLContextUPtr = std::unique_ptr<
+    _cl_context, decltype(ContextDeleter)>; // decltype(CLContextUPtr.get()) is
+                                            // same as cl_context, should
+                                            // initialized with ContextDeleter
+
+/*! \fn std::tuple<std::vector<CLProgramUPtr>, std::string, cl_int>
```

Comment:
Is it really required to duplicate function prototype in this comment? It doesn't make a lot of sense, since it is available below

Reply:
Fixed